rust/rope/src/find.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of string finding in ropes.
+
+use rope::RopeInfo;
+use tree::Cursor;
+
+use memchr::{memchr, memchr2, memchr3};
+
+pub enum FindResult {
+    Found(usize),
+    NotFound,
+    TryAgain,
+}
+
+/// Case-sensitive find (just does byte comparison, no Unicode logic).
+///
+/// Can panic if `pat` is empty.
+pub fn find(cursor: &mut Cursor<RopeInfo>, pat: &str) -> Option<usize> {
+    loop {
+        match find_step(cursor, pat) {
+            FindResult::Found(pos) => return Some(pos),
+            FindResult::NotFound => return None,
+            _ => (),
```

Comment:
given that this is a single case, handling it explicitly would make the control flow clearer.